featurebyte/query_graph/transform/null_filling_value.py

```diff
@@ -0,0 +1,157 @@
+            # for aggregation node, we replace it with a request column node
+            op_struct = global_state.operation_structure_map[node.name]
+            node_to_insert = RequestColumnNode(
+                name="",
+                parameters=RequestColumnNode.RequestColumnNodeParameters(
+                    column_name="agg_col", dtype=op_struct.aggregations[0].dtype
+                ),
+                output_type=NodeOutputType.SERIES,
+            )
+
+        inserted_node = global_state.graph.add_operation_node(
+            node=node_to_insert, input_nodes=mapped_input_nodes
+        )
+
+        # update the node name map
+        global_state.node_name_map[node.name] = inserted_node.name
+        return global_state
+
+    def extract(self, node: Node, **kwargs: Any) -> NullFillingValueGlobalState:
+        # extract operation structure of the graph that used to identify the dtype of the aggregation
+        op_struct_info = OperationStructureExtractor(graph=self.graph).extract(node=node)
+
+        # construct query graph for constructing null filling value extraction function
+        state: NullFillingValueGlobalState = self._extract(
+            node=node,
+            branch_state=BaseModel(),
+            global_state=NullFillingValueGlobalState(
+                operation_structure_map=op_struct_info.operation_structure_map
+            ),
+            topological_order_map=self.graph.node_topological_order_map,
+        )
+
+        # convert query graph to actual python code
+        mapped_node = state.graph.get_node_by_name(state.node_name_map[node.name])
+        code_state = OnDemandFeatureFunctionExtractor(graph=state.graph).extract(
+            node=mapped_node,
+            sql_function_name="",
+            sql_input_var_prefix="",
+            sql_request_input_var_prefix="",
+            sql_comment="",
+            function_name="extract_null_filling_value",
+            input_var_prefix="",
+            request_input_var_prefix="agg_col",
+            output_dtype=DBVarType.FLOAT,
+        )
+        input_values = []
+        for sql_input_info in code_state.sql_inputs_info:
+            val = "np.nan"
+            if sql_input_info.py_type == "pd.Timestamp":
+                val = "pd.NaT"
+            input_values.append(f"{sql_input_info.py_input_var_name}={val}")
+
+        null_filling_value_codes = code_state.generate_code(to_sql=False)
+
+        # inject null value to the function to extract null filling value
+        input_params = ", ".join(input_values)
+        null_filling_value_codes += f"\n\nfill_value = extract_null_filling_value({input_params})\n"
+        state.codes = null_filling_value_codes
+        scope: Dict[str, Any] = {}
+        exec(null_filling_value_codes, scope)  # pylint: disable=exec-used  # nosec
+        fill_value = scope["fill_value"]
+
+        # if fill_value is not null, that implies the query graph has null filling operation
```

Comment:
If fill value is na should we skip the on-demand feature view altogether?

Reply:
Nope, this class is used to extract fillna value, it is not used to construct on-demand feature view directly.